storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java

```diff
@@ -29,25 +35,768 @@
+                            else if (segment.baseOffset() == maxOffsetMetadata.segmentBaseOffset && !maxOffsetMetadata.messageOffsetOnly())
+                                maxPositionOpt = Optional.of((long) maxOffsetMetadata.relativePositionInSegment);
+                            else
+                                maxPositionOpt = Optional.empty();
+
+                            fetchDataInfo = segment.read(startOffset, maxLength, maxPositionOpt, minOneMessage);
+                            if (fetchDataInfo != null) {
+                                if (includeAbortedTxns) {
+                                    fetchDataInfo = addAbortedTransactions(startOffset, segment, fetchDataInfo);
+                                }
+                            } else {
+                                segmentOpt = segments.higherSegment(baseOffset);
+                            }
+                        }
+
+                        if (fetchDataInfo != null) {
+                            return fetchDataInfo;
+                        } else {
+                            // okay we are beyond the end of the last segment with no data fetched although the start offset is in range,
+                            // this can happen when all messages with offset larger than start offsets have been deleted.
+                            // In this case, we will return the empty set with log end offset metadata
+                            return new FetchDataInfo(nextOffsetMetadata, MemoryRecords.EMPTY);
+                        }
+                    }
+                }
+        );
+    }
+
+    public void append(long lastOffset, long largestTimestamp, long shallowOffsetOfMaxTimestamp, MemoryRecords records) throws IOException {
+        segments.activeSegment().append(lastOffset, largestTimestamp, shallowOffsetOfMaxTimestamp, records);
+        updateLogEndOffset(lastOffset + 1);
+    }
+
+    FetchDataInfo addAbortedTransactions(long startOffset, LogSegment segment, FetchDataInfo fetchInfo) throws IOException {
+        int fetchSize = fetchInfo.records.sizeInBytes();
+        OffsetPosition startOffsetPosition = new OffsetPosition(
+                fetchInfo.fetchOffsetMetadata.messageOffset,
+                fetchInfo.fetchOffsetMetadata.relativePositionInSegment);
+        long upperBoundOffset = segment.fetchUpperBoundOffset(startOffsetPosition, fetchSize)
+                .orElse(segments.higherSegment(segment.baseOffset())
+                        .map(LogSegment::baseOffset).orElse(logEndOffset()));
+
+        List<FetchResponseData.AbortedTransaction> abortedTransactions = new ArrayList<>();
+        Consumer<List<AbortedTxn>> accumulator = abortedTxns -> {
+            for (AbortedTxn abortedTxn : abortedTxns)
+                abortedTransactions.add(abortedTxn.asAbortedTransaction());
+        };
+        collectAbortedTransactions(startOffset, upperBoundOffset, segment, accumulator);
```

Comment:
not sure why we need to use `accumulator` pattern ... could you please make `collectAbortedTransactions` return `abortedTransactions` directly? for example:

```java
        var abortedTransactions = collectAbortedTransactions(startOffset, upperBoundOffset, segment)
                .stream().map(AbortedTxn::asAbortedTransaction).toList();
        return new FetchDataInfo(fetchInfo.fetchOffsetMetadata,
                fetchInfo.records,
                fetchInfo.firstEntryIncomplete,
                Optional.of(abortedTransactions));
    }

    private List<AbortedTxn> collectAbortedTransactions(long startOffset, long upperBoundOffset,
                                                                                  LogSegment startingSegment) {
```

I suppose the downside is the extra loop, but I don't think that would lead to significant performance regression...

Reply:
I'm uneasy trying to include many refactors with the rewrite. I'd rather make changes like this in a follow up PR.